addons/web/static/src/scss/bootstrap_review_frontend.scss

```diff
@@ -15,27 +15,6 @@
     @include text-emphasis-variant(".text-#{$color}", $value);
```

Comment:
Here are the corrections I made (to make you gain some time after my long-awaited review :smiling_face_with_tear: )
Note that I am putting the whole files, as it is easier for me :wink: (you will see the diff when pasting the content in your editor :slightly_smiling_face:)

I added review comments on the commit so read them first (to understand what was wrong) and then you can use my code to fix the issues :slightly_smiling_face: (if you agree with my fixes of course) 
Note that I advise you to double check the behaviors (there is one that is weird but it was also the case before my changes (there is a "Beahviour" review comment about that) :+1: 

The fixes are mainly to:
- correctly use the options (so `computeWidgetVisibility`, `data-name`,...)
- removing the use of jQuery, 
- use double quotes in string,
- renaming some function into names I think are better (personal opinion, feel free to change them)
- + some website conventions

Reply:
`s_card.xml` file (note that I did not manage the version attributes (`data-vxml/vcss`):

```xml
<?xml version="1.0" encoding="utf-8"?>
<odoo>

<template id="s_card" name="Card">
    <div class="s_card o_card_img_top card o_cc o_cc1">
        <figure class="o_card_img_wrapper ratio ratio-16x9 mb-0">
            <img class="o_card_img card-img-top object-fit-cover" src="/web/image/website.s_card_default_image_1"/>
        </figure>
        <div class="card-body">
            <h5 class="card-title">Card title</h5>
            <p class="card-text">A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety of content, contextual background colors, and powerful display options.</p>
            <a t-att-href="cta_btn_href">Read More</a>
        </div>
    </div>
</template>

<template id="s_card_options" inherit_id="website.snippet_options">
    <xpath expr="." position="inside">
        <t t-call="website.snippet_options_background_options">
            <t t-set="selector" t-value="'.s_card'"/>
            <t t-set="with_colors" t-value="True"/>
            <t t-set="with_images" t-value="True"/>
            <t t-set="with_shapes" t-value="True"/>
            <t t-set="with_color_combinations" t-value="True"/>
            <t t-set="with_gradients" t-value="True"/>
        </t>

        <div data-js="Box" data-selector=".s_card">
            <t t-call="website.snippet_options_border_widgets"/>
            <t t-call="website.snippet_options_shadow_widgets"/>
        </div>

        <div data-js="CardWidth" data-selector=".s_card">
            <we-range string="Card Width"
                data-select-style="0"
                data-css-property="max-width"
                data-unit="%"
                data-min="8"
                data-max="100"
                data-display-range-value="true"
                data-display-range-value-unit="%"/>
        </div>

        <div data-js="CardImageOptions" data-selector=".s_card">
            <we-row string="Cover Image">
                <we-button title="Add" data-add-cover-image="" data-no-preview="true" class="o_we_bg_success" data-name="add_cover_image_opt">Add</we-button>
            </we-row>

            <we-row string="Cover Image">
                <we-button-group data-name="cover_image_position_opt">
                    <we-button title="Top"
                        data-name="card_img_top_opt"
                        data-select-class="o_card_img_top"
                        data-select-image-class="card-img-top"
                        data-img="/website/static/src/img/snippets_options/pos_top.svg"/>
                    <we-button title="Left"
                        data-select-class="o_card_img_horizontal flex-sm-row"
                        data-select-image-class="rounded-start"
                        data-img="/website/static/src/img/snippets_options/pos_left.svg"/>
                    <we-button title="Right"
                        data-select-class="o_card_img_horizontal flex-sm-row-reverse"
                        data-select-image-class="rounded-end"
                        data-img="/website/static/src/img/snippets_options/pos_right.svg"/>
                </we-button-group>
                <we-button title="Remove Cover" data-remove-cover-image="" data-no-preview="true" class="fa fa-trash-o fa-fw o_we_bg_danger" data-name="remove_cover_image_opt"/>
            </we-row>

            <we-select string="Ratio" data-apply-to=".o_card_img_wrapper" class="o_we_sublevel_1">
                <we-button data-select-class="">Image default</we-button>
                <we-button data-select-class="ratio ratio-1x1">Square</we-button>
                <we-button data-select-class="ratio ratio-4x3" data-dependencies="card_img_top_opt">Landscape - 4/3</we-button>
                <we-button data-select-class="ratio ratio-16x9" data-dependencies="card_img_top_opt">Wide - 16/9</we-button>
                <we-button data-select-class="ratio ratio-21x9" data-dependencies="card_img_top_opt">Ultrawide - 21/9</we-button>
                <we-button data-select-class="ratio o_card_img_ratio_custom" data-dependencies="card_img_top_opt" data-name="image_ratio_custom_opt">Custom</we-button>
            </we-select>

            <!-- Get bootstrap default column size -->
            <t t-set="colSize" t-value="100/12"/>
            <we-range string="Width" class="o_we_sublevel_1"
                title="Adjust the image width"
                data-name="cover_image_width_opt"
                data-dependencies="!card_img_top_opt"
                data-select-style="0"
                data-css-property="--card-img-size-h"
                data-unit="%"
                t-att-data-min="colSize"
                t-att-data-max="colSize * 11"
                t-att-data-step="colSize"/>

            <we-range
                string="Custom Ratio" class="o_we_sublevel_2"
                data-dependencies="image_ratio_custom_opt"
                data-select-style="0"
                data-css-property="--card-img-aspect-ratio"
                data-unit="%"
                data-display-range-value="true"
                data-display-range-value-unit="%"
                data-min="8"
                data-max="132"
                data-step="4"/>

            <we-range
                string="Alignment" class="o_we_sublevel_1"
                data-name="cover_image_alignment_opt"
                data-select-style="0"
                data-align-cover-image=""
                data-css-property="--card-img-ratio-align"
                data-unit="%"
                data-min="0"
                data-max="100"/>
        </div>
    </xpath>
</template>

<record id="website.s_card_000_scss" model="ir.asset">
    <field name="name">Card 000 SCSS</field>
    <field name="bundle">web.assets_frontend</field>
    <field name="path">website/static/src/snippets/s_card/000.scss</field>
</record>


</odoo>

```

Reply:
`options.js` file:
```js
/** @odoo-module **/

import { renderToElement } from "@web/core/utils/render";
import options from "@web_editor/js/editor/snippets.options";


options.registry.CardWidth = options.Class.extend({
    /**
     * @override
     */
    _computeWidgetState(methodName, params) {
        if (methodName === "selectStyle") {
            if (params.cssProperty === "max-width") {
                // If no `max-width` is set, consider it to be at 100%.
                if (!this.$target[0].style[params.cssProperty]) {
                    return "100%";
                }
            }
        }
        return this._super(...arguments);
    },
});

options.registry.CardImageOptions = options.Class.extend({
    /**
     * @override
     */
    start() {
        this._computeAlignmentDirection();
    },

    //--------------------------------------------------------------------------
    // Options
    //--------------------------------------------------------------------------

    /**
     * Injects a new cover image.
     */
    addCoverImage() {
        const imageWrapperEl = renderToElement("website.s_card.imageWrapper");
        this.$target[0].insertAdjacentElement("afterbegin", imageWrapperEl);
    },
    /**
     * Changes the cover image position.
     *
     * @see this.selectClass for parameters
     */
    selectImageClass(previewMode, widgetValue, params) {
        const imageEl = this.$target[0].querySelector(".o_card_img");
        for (const className of params.possibleValues) {
            if (className) {
                imageEl.classList.remove(className);
            }
        }
        imageEl.classList.add(widgetValue);
    },
    /**
     * Removes the cover image.
     */
    removeCoverImage() {
        const imageWrapperEl = this.$target[0].querySelector(".o_card_img_wrapper");
        imageWrapperEl.remove();
    },
    /**
     * Aligns the image inside the cover.
     */
    alignCoverImage() {
        this._computeAlignmentDirection();
    },

    //--------------------------------------------------------------------------
    // Private
    //--------------------------------------------------------------------------

    /**
     * @override
     */
    _computeWidgetVisibility(widgetName, params) {
        const hasCoverImage = !!this.$target[0].querySelector(".o_card_img_wrapper");
        if (widgetName === "add_cover_image_opt") {
            return !hasCoverImage;
        } else if (["cover_image_position_opt", "remove_cover_image_opt", "cover_image_width_opt"].includes(widgetName)) {
            return hasCoverImage;
        } else if (widgetName === "cover_image_alignment_opt") {
            const hasNonSquareRatio = !!this.$target[0].querySelector(".o_card_img_wrapper.ratio:not(.ratio-1x1)");
            return hasCoverImage && hasNonSquareRatio;
        }
        return this._super(...arguments);
    },
    /**
     * Changes the alignment widget behavior according to the image ratio.
     */
    _computeAlignmentDirection() {
        const imageEl = this.$target[0].querySelector(".o_card_img");
        const imageWrapperEl = this.$target[0].querySelector(".o_card_img_wrapper");
        if (!imageEl || !imageWrapperEl) {
            return;
        }

        const imgRatio = imageEl.offsetHeight / imageEl.offsetWidth;
        const wrapperRatio = imageWrapperEl.offsetHeight / imageWrapperEl.offsetHeight;
        const alignVertically = imgRatio <= wrapperRatio;
        this.$target[0].classList.toggle("o_card_img_adjust_v", alignVertically);
        this.$target[0].classList.toggle("o_card_img_adjust_h", !alignVertically);
    },
});

```

Reply:
Unfortunately something was off compared to my initial implementation.
The custom ratio slider was visible for the horizontal layout as well. I addressed the issue in JS, let me know what you think!

---

Your poposal

https://github.com/user-attachments/assets/1340d946-8e30-4f1f-9789-a1210e41b72a

---

Old implementation 

https://github.com/user-attachments/assets/1bf4a22e-1f19-4f0c-81ac-45371856467d

Reply:
Yes I noticed it, it happens because of the "weird behavior" I mentioned [here](https://github.com/odoo/odoo/pull/172449#discussion_r1686927475) :wink:  
=> If the invalid ratio classes (so the ones that should not appear when the image is horizontal) are not removed when changing the option, the "Custom" value is still selected, and because of that, the "Custom Ratio" slider is still displayed (because of the dependencies) 
That's why I didn't do anything to hide it, because I first wanted to know if keeping "Custom" selected shouldn't be fixed first (if it is, the slider will automatically be hidden with it) :slightly_smiling_face: